crates/ruff_python_formatter/src/comments/visitor.rs

```diff
@@ -0,0 +1,354 @@
+            walk_expr(self, format_spec);
+        }
+        self.finish_node(format_spec);
+    }
+
+    fn visit_arguments(&mut self, arguments: &'ast Arguments) {
+        if self.start_node(arguments).is_traverse() {
+            walk_arguments(self, arguments);
+        }
+        self.finish_node(arguments);
+    }
+
+    fn visit_arg(&mut self, arg: &'ast Arg) {
+        if self.start_node(arg).is_traverse() {
+            walk_arg(self, arg);
+        }
+        self.finish_node(arg);
+    }
+
+    fn visit_keyword(&mut self, keyword: &'ast Keyword) {
+        if self.start_node(keyword).is_traverse() {
+            walk_keyword(self, keyword);
+        }
+        self.finish_node(keyword);
+    }
+
+    fn visit_alias(&mut self, alias: &'ast Alias) {
+        if self.start_node(alias).is_traverse() {
+            walk_alias(self, alias);
+        }
+        self.finish_node(alias);
+    }
+
+    fn visit_withitem(&mut self, withitem: &'ast Withitem) {
+        if self.start_node(withitem).is_traverse() {
+            walk_withitem(self, withitem);
+        }
+
+        self.finish_node(withitem);
+    }
+    fn visit_match_case(&mut self, match_case: &'ast MatchCase) {
+        if self.start_node(match_case).is_traverse() {
+            walk_match_case(self, match_case);
+        }
+        self.finish_node(match_case);
+    }
+
+    fn visit_pattern(&mut self, pattern: &'ast Pattern) {
+        if self.start_node(pattern).is_traverse() {
+            walk_pattern(self, pattern);
+        }
+        self.finish_node(pattern);
+    }
+}
+
+fn text_position(comment_range: TextRange, source_code: SourceCode) -> CommentTextPosition {
+    let before = &source_code.as_str()[TextRange::up_to(comment_range.start())];
+
+    for c in before.chars().rev() {
+        match c {
+            '\n' | '\r' => {
+                break;
+            }
+            c if is_python_whitespace(c) => continue,
+            _ => return CommentTextPosition::EndOfLine,
+        }
+    }
+
+    CommentTextPosition::OwnLine
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+enum TraversalSignal {
+    Traverse,
+    Skip,
+}
+
+impl TraversalSignal {
+    const fn is_traverse(self) -> bool {
+        matches!(self, TraversalSignal::Traverse)
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+struct CommentsBuilder<'a> {
+    comments: CommentsMap<'a>,
+}
+
+impl<'a> CommentsBuilder<'a> {
+    fn add_comment(&mut self, placement: CommentPlacement<'a>) {
+        match placement {
+            CommentPlacement::Leading { node, comment } => {
+                self.push_leading_comment(node, comment);
+            }
+            CommentPlacement::Trailing { node, comment } => {
+                self.push_trailing_comment(node, comment);
+            }
+            CommentPlacement::Dangling { node, comment } => {
+                self.push_dangling_comment(node, comment);
+            }
+            CommentPlacement::Default(comment) => {
```

Comment:
I'm having trouble following `CommentPlacement::Default`. `place_comment` always returns default, so we're always taking this path. Why does `CommentPlacement::Default` exist at all? Could `place_comment` not contain this logic?

Reply:
Oh, sorry, there's a comment about this but it's marked as "outdated" so didn't appear in the diff. Reading...

Reply:
I think my question still stands -- why does `Default` exist? Can `place_comment` not contain this logic + any special-casing?

Reply:
`place_comment` could contain all the logic, but I prefer to have it separated because the custom logic itself can get very messy ([see rome](https://github.com/rome/tools/blob/main/crates/rome_js_formatter/src/comments.rs#L156-L1165)). Other considerations are:

* At least for rome, the `Default` logic was shared across all languages, and `place_comment` was language specific. This does not (or not yet?) apply to ruff. 
* The way `place_comment` works is to have different handle functions that pass the comment through, for as long as the prevous handler returned `CommentPlacement::Default` (chain of responsibility pattern?). Option doesn't work because we need to pass `comment` as an owned value and returning `None` would "drop" the comment and we could not retrieve it again. We could change the design to decouple the "placement" and the comment, pass the comment by reference, and change the return type of `place_comment` to `(Placement, Comment)`. Individual handlers could then return a `Option<Placement>`. The downside of this is that it becomes easier to "ignore" the value of a handler because the `comment `doesn't get "consumed" when explicitly specifying that it's a *leading*, *dangling* or *trailing* comment (you can still manually extract it again, but you would need to be very deliberate about it). 
* Having `CommentPlacement::Default` as a concept provides a great opportunity to document the rule with examples. 

Would you find it easier to understand if all the logic is in one place? Where would you document the "default" behavior?

Reply:
Seeing ahead to the version in Rome is helpful. I think the thing I was missing here was that I was wondering: won't this restrict our ability to use `CommentPlacement` elsewhere? E.g., imagine that we wanted to use `CommentPlacement` in the formatter itself, to figure out whether a given comment should be printed before or after a given node. In that case... how would we handle `CommentPlacement::Default`?

But, the comments data structure effectively implements a refined enum by way of tracking leading, trailing, and dangling comments. So this `CommentPlacement` enum is only used in the building phase, and `CommentPlacement::Default` doesn't extend outwards to any sites in which it's existence would be confusing or cause problems.

Reply:
That's a good point. I think I'll move the definition into the visitor.rs file to make that clear. The placement.rs will only contain our custom logic